### PR TITLE
[WIP] Bugzilla

### PR DESCRIPTION
**Pull request by vaartis**
###### Motivation for this change

There was no bugzilla expression
So.. This is my first contribution and i am not quite sure if everything is good, there probably should be some kind of systemd service to launch bugzilla but i don't really know how to write them. There are also some of bugzilla's CPAN dependencies. Please, help me on systemd service and finilizing the package

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by vaartis** [comment]
Perl packages can be merged, i think, should i put them in a separate PR?
/cc @joachifm

**Comment by joachifm** [comment]
If the packages are useful on their own, I think it makes sense to merge them before the systemd service is ready.

I don't see much to nitpick about here, except maybe providing proper licenses for all packages or at least indicating whether they're free software.

@rycee may be able to help more with the Perl stuff.

**Comment by qknight** [comment]
@joachifm 
i think we should allow `unknown` in principle but in an extra commit add it to `lib/licenses.nix` and 
```
  unknown = {
    fullName = "unknown";
    free = false;
  };
```
what do you think?

@vaartis you can simply put all the perlPackages into one PR. that said, i would even squash this whole PR into one commit.

**Comment by vaartis** [comment]
@joachifm this licenses are taken from CPAN, some of them are marked as unknown on cpan

**Comment by rycee** [comment]
Unfortunately many CPAN authors don't set correct meta information for the license but usually you can find it out by browsing the package a bit.

In particular
* For `AuthenRaidus` please set `artistic2` as license: https://st.aticpan.org/source/PORTAONE/Authen-Radius-0.26/LICENSE
* For `Chart` please set `[ artistic1 gpl1Plus ]` as license: https://metacpan.org/pod/distribution/Chart/Chart.pod#COPYRIGHT
* For `PatchReader` please set `artistic1` as license: https://github.com/tmannerm/PatchReader#readme
* And for `TemplateGD` please set `[ artistic1 gpl1Plus ]` as license: https://metacpan.org/pod/Template::Plugin::GD#COPYRIGHT

About squashing I personally prefer one commit per `init` but I have no strong feelings about it. I'm fine with having something like `perlPackages: init several packages` as the commit header and have individual lines in the commit message body, e.g., `EmailReply at 1.204`.

**Comment by vaartis** [comment]
Perl modules PR: #26569

**Comment by vaartis** [comment]
@joachifm so.. i've kind of written the service, but how do i test it? Obviously, it isn't available from system-wide `configuration.nix`, so how do i start it then?

**Comment by joachifm** [comment]
@vaartis you can add the module to https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix or add it to the `imports` in your configuration.nix. Then you can use it like any other module.

**Comment by joachifm** [comment]
Or you can define the service inline in configuration.nix, as in `systemd.services.bugzilla = { ... }`.

**Comment by rasendubi** [comment]
> Also, you can't really close the source of perl programms, can you?

Licenses are not really about closing sources. You can publish your source publicly still forbid anyone to use it; on the other hand, you can do open source license with closed code: you only need to provide source code when you sell the software.

To the topic... the PR looks good superficially, though indentation level is 4 spaces when most of nixpks use 2.

For systemd module, check out [nixos/modules/services](https://github.com/NixOS/nixpkgs/tree/master/nixos/modules/services) directory—it has plenty of examples. You should grasp the pattern quickly.

**Comment by vaartis** [comment]
Thanks! Also, aparently there is an error in httpd module, which breaks mod_perl. Somebody moved httpd modules into `pkgs.apacheHttpdPackagesFor`, but did not update httpd package, it seems

**Comment by vaartis** [comment]
`The option `services.httpd.adminAddr' is used but not defined.`
Well, i don't even know what to do with this.. It really isn't defined in httpd, but used in several other expressions

**Comment by qknight** [comment]
the option `adminAddr` is enforced so that you pass in an email address.

are you packaging `bugzilla` as a `submodule` for `httpd`? have a look at this mediawiki example:

https://web.archive.org/web/20160829175239/https://nixos.org/wiki/MediaWiki

**Comment by vaartis** [comment]
Hmm.. Is there any way to test httpd submodule without rebuilding whole system to the latest version from github (used it by setting `NIX_PATH` to checked out dir). Just importing bugzilla httpd submodule says that httpd's `documentRoot` and other does not exist..

**Comment by qknight** [comment]
# build the system

    nixos-rebuild -I nixpkgs=../mymodified/nixpkgs/ switch

is the command you can use to build your new webserver. if you want to test on master, then you probably have to rebuild much software but only commits on `master` are valuable, since we always commit to master and not 17.03 (or similar).

to deveop a feature it might be a good idea to use 17.03 as basis (or whatever you are working on, see your nix-channel).

**Comment by vaartis** [comment]
I don't think `nixos-rebuild -I "nixpgs=/home/vaartis/Code/nixpkgs/" test` is working.. It just seems like it ignores the path and uses the system one
```
error: getting status of ‘/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs/nixos/modules/services/web-servers/apache-httpd/bugzilla.nix’: No such file or director
```
So, it doesn't use my path, despite apache being there

**Comment by bjornfor** [comment]
> I don't think nixos-rebuild -I nixpgs=/home/vaartis/Code/nixpkgs/ test is working..

Typo: nixpgs -> nixpkgs

File: pkgs/servers/bugzilla/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, pkgs, fetchurl, perlPackages }:
+
+stdenv.mkDerivation rec {
+
+    name = "bugzilla-5.1.1";
+
+    src = fetchurl {
+        url = "https://ftp.mozilla.org/pub/webtools/${name}.tar.gz";
```
Comment (garbas) [inline pkgs/servers/bugzilla/default.nix]: maybe it would be better to fetch from github directly https://github.com/bugzilla/bugzilla

**Review by garbas — APPROVED** [review]
@vaartis again, thank you.

💯 from my side. Once we have it packaged we can look how to generate Perl expressions via carton or such Perl packaging tools.

I'm approving this ahead of time, since I wont have time in coming days.

**Comment by vaartis** [comment]
@bjornfor oh... My bad

**Comment by vaartis** [comment]
So.. bugzilla requires the configure script to be ran by somebody in `wwwrun` group, but `nixbld` user is not in that group, how can i add it to group?

**Comment by vaartis** [comment]
So, i've just pushed everything i could do with it, probably cannot continue it until i figure out how to add modules to apache perl search path
```
Jun 17 17:49:53 Nyak httpd[16207]: AH00526: Syntax error on line 168 of /nix/store/kmjqycgwac84hdvqda4k49ji025qgrnz-httpd.conf:
Jun 17 17:49:53 Nyak httpd[16207]: Can't locate Apache2/Log.pm in @INC
```

**Comment by 0xABAB** [comment]
@vaartis You want the `-I` option
`  PerlSwitches -I/home/httpd/project2/lib` (see: https://perl.apache.org/docs/2.0/user/config/config.html )

**Comment by 0xABAB** [comment]
See https://github.com/NixOS/nixpkgs/pull/26978 for an example.

**Comment by vaartis** [comment]
@0xABAB can't work on this right now, feel free to copy my working branch and finish this one..

**Comment by aanderse** [comment]
Is there interest in continuing this PR from any involved parties?

**Comment by vaartis** [comment]
I'm probably not going to finish it.

**Pull request closed without merging**